adobe/spectrum-web-components
# build: migrate to changesets from lerna

## Description

This removes [Lerna](https://lerna.js.org/) and replaces it with [Changesets](https://github.com/changesets/changesets) for handling package versioning and publishing (releases). Changesets also respect the files field in package.json that explicitly states which files should be included in the publish. This is not considered by lerna.


## How and where has this been tested?

Please tag yourself on the tests you've marked complete to confirm the tests have been run by someone other than the author.

### Validation steps


1. Please read the [Intro to Using Changesets](https://github.com/changesets/changesets/blob/main/docs/intro-to-using-changesets.md) documentation.
2. Please read the description of the [Changesets GitHub bot](https://github.com/apps/changeset-bot).
3. Please read the [documentation on adding a changeset](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md).
4. Finally, this [blog post](https://brionmario.medium.com/changesets-is-a-game-changer-fe752af6a8cc) is handy and gives more information about using Changesets.
5. I added two commits (a change to Well, and a changeset) to demonstrate what this would look like when the GitHub bot detected a changeset. I'll drop those two commits before merging.

It's important to know that moving to Changesets shifts the onus of version determination toward a more intentional choice for contributors. Some could argue that the onus was already on the contributor as we used Conventional Commit messages to provide Lerna a way to infer the severity of a version increase. Now, instead of relying on Conventional Commit messages, the contributor will be asked to [provide changesets](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md#i-am-in-a-multi-package-repository-a-mono-repo) as part of their PR process.

The workflow for this will look something like:
1. The contributor makes their changes and commits them.
2. Locally, the contributor runs `yarn changeset` and is asked in the CLI to choose which package(s) should be part of the changeset.
3. After the package(s) have been selected, the CLI will ask if the version increment should be a `major`, `minor`, or `patch`. Hitting `enter` in the CLI without making a choice will skip options with `patch` being the final option.
4. Next, the CLI will ask for a summary of the changes. This summary will be used in the `changelog.md` for the respective package(s). We're also using the `@changesets/changelog-github` package to provide additional GitHub-related context (pull request number + link, contributor information), and this info will show in the `changelog.md` for the package(s), as well.
7. Stage the change, commit the change, and push to the remote branch.

Alternatively, After committing all the changes, the contributor can also leverage changeset bot to add the chanesets.
This bot will comment on PRs saying that either a user might need to add a changeset(note that PRs changing things like documentation generally don't need a changeset)or say that the PR is good and already has a changeset.
For more details: https://github.com/changesets/bot

## Motivation and context



## How has this been tested?




    1. Go here
    2. Do this
    1. Go here
    2. Do this

## Screenshots (if appropriate)

## Types of changes




## Checklist




## Best practices

This repository uses conventional commit syntax for each commit message; note that the GitHub UI does not use this by default so be cautious when accepting suggested changes. Avoid the "Update branch" button on the pull request and opt instead for rebasing your branch against `main`.
